### PR TITLE
Optimize writes: Use Telegraf to pre-process data before writing

### LINKED ISSUE #5155: Telegraf guide: how to use the Converter processor to coerce data before writing
-  how to use the Converter processor to coerce data before writing
- how to use Exec and Grok to filter data before writing.

<@U01TCPBF3U4> could you please add this to the docs backlog, if you don't already have something?
&gt;  That sounds like an opportunity for a KB article, etc to get in front of any issue and turn it into a productive ehancement.

[Slack Message](https://influxdata.slack.com/archives/C04768675QD/p1695828045290899?thread_ts=1695822639.441929&cid=C04768675QD)

### PR DESCRIPTION
**Pull request by jstirnaman**
Closes #5155 

- Add Telegraf examples that pre-process data to avoid schema conflicts and reduce payload (bandwith, memory, etc.) when writing data.

- Includes Dedup
- Passes tests

### CODE REVIEW

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 1]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
 ## Write multiple data points in one request
 
 To write multiple lines in one request, each line of line protocol must be delimited by a new line (`\n`).
+
+## Pre-process data before writing
+
+Pre-processing data in your write workload can help you avoid [write failures](/influxdb/cloud-serverless/write-data/troubleshoot/#troubleshoot-failures) due to schema conflicts or resource use.
+For example, if you have many devices that write to the same measurement, and some devices use different data types for the same field, then you might want to generate an alert or convert field data to fit your schema before you send the data to InfluxDB.
+
+With Telegraf, you can process data from other services and files and then write it to InfluxDB.
```
Comment (sanderson) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:152]:
```suggestion
With [Telegraf](/telegraf/v1/), you can process data from other services and files and then write it to InfluxDB.
```

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 2]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+In addition to processing data with Telegraf's included plugins, you can use the [Execd processor plugin](/telegraf/v1/plugins/#processor-execd) to integrate your own code and external applications.
+
+The following examples show how to [configure](/telegraf/v1/configuration) the Telegraf agent and [plugins](/telegraf/v1/plugins/) to optimize writes.
```
Comment (sanderson) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:155]: For SEO

```suggestion
The following examples show how to [configure the Telegraf agent](/telegraf/v1/configuration) and [plugins](/telegraf/v1/plugins/) to optimize writes.
```

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 3]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+    tagexclude = ["host"]
+  [[outputs.influxdb_v2]]
+    urls = ["HOST"]
+    token = "API_TOKEN"
+    organization = ""
+    bucket = "BUCKET_NAME"
+EOF
+```
+
+Replace the following:
```
Comment (sanderson) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:184]: Missing placeholder values

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 4]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+    fieldpass = ["usage_system", "usage_idle"]
+    # Remove the specified tags from points.
+    tagexclude = ["host"]
+  [[outputs.influxdb_v2]]
+    urls = ["HOST"]
+    token = "API_TOKEN"
+    organization = ""
+    bucket = "BUCKET_NAME"
+EOF
+```
```
Comment (sanderson) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:182]:
~~~suggestion
```sh
cat <<EOF >> ./telegraf.conf
  [[inputs.cpu]]
    # Remove the specified fields from points.
    fieldpass = ["usage_system", "usage_idle"]
    # Remove the specified tags from points.
    tagexclude = ["host"]
  [[outputs.influxdb_v2]]
    urls = ["{{< influxdb/host >}}"]
    token = "API_TOKEN"
    organization = ""
    bucket = "BUCKET_NAME"
EOF
```
~~~

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 5]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+For example, if you write the sample data in
+[Get started home sensor data](/influxdb/cloud-serverless/reference/sample-data/#get-started-home-sensor-data) to a bucket, and then try to write the following batch to the same measurement:
+
+```text
+home,room=Kitchen temp=23.1,hum=36.6,co=22.1 1641063600
+home,room=Living\ Room temp=22i,hum=36.4,co=17i 1641067200
+home,room=Kitchen temp=22.7,hum=36.5,co=26i 1641067200
+```
+
+InfluxDB expects `co` to contain an integer and rejects the point in which `co` contains a floating-point decimal (`22.1`).
```
Comment (sanderson) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:212]:
```suggestion
InfluxDB expects `co` to contain an integer value and rejects points with `co` floating-point decimal (`22.1`) values.
```

**Review by sanderson — CHANGES_REQUESTED** [review]
Just some minor suggestions and missing placeholder descriptions.

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 6]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+   <!--pytest-codeblocks:expected-output-->
+
+   ```
+   > home,room=Kitchen co=22i,hum=36.6,temp=23.1 1641063600000000000
+   > home,room=Living\ Room co=17i,hum=36.4,temp=22.7 1641063600000000000
+   ```
+
+<!-- ### Avoid sending duplicate data
+  -- Can't get processors.dedup to work as expected.
+  -- The content for this section is in branch https://github.com/influxdata/docs-v2/compare/issue5155-preprocess-with-dedup
```
Comment (powersj) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:385]: I responded to your question in slack, but it was due to the timestamps you have on the data. The dedup processor takes into account the timestamp given the dedeup interval.

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 7]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+-->
+
+### Run custom preprocessing code
+
+Use Telegraf and the [Execd processor plugin](/telegraf/v1/plugins/#processor-execd) to execute code external to Telegraf and then write the processed data.
```
Comment (powersj) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:390]: Is there going to be an example here like the others?

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md (hunk 3 above)
Comment (jstirnaman) [reply content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:184]: 🤦

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 8]
```diff
@@ -122,3 +143,248 @@ If a host's clocks isn't synchronized with NTP, timestamps may be inaccurate.
...
+   home,room=Living\ Room co=17i 1641063600
+   EOF
+   grace_seconds=$(($(date +%s)-1641063000))
+   grace="${grace_seconds}s"
+   ```
+
+2. Enter the following command to configure Telegraf to parse the file, merge the points, and write the data to a bucket. To merge series, you must specify the following in your Telegraf configuration:
+
+   - the timestamp precision for your input data (for example, `influx_timestamp_precision` for line protocol)
+   - Optional: `aggregators.merge.grace` to extend the window and allow more points to be merged. For demonstration purposes, the following example sets `grace` to a large duration to include the sample data timestamps.
```
Comment (jstirnaman) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:321]:
```suggestion
   - Optional: `aggregators.merge.grace` to extend the window and allow more points to be merged. For demonstration purposes, the following example sets `grace` to a large duration that includes the sample data timestamps.
```

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md (hunk 7 above)
Comment (jstirnaman) [reply content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:390]: > Is there going to be an example here like the others?

Yes, but I realized I didn't have a good write example at hand. Can you point me to one?

Comment (powersj) [reply content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:390]: Essentially the [execd pluin](https://github.com/influxdata/telegraf/tree/master/plugins/processors/execd) will  send the user line protocol on stdin and expect line protocol back out on stdout. There is a go example in the plugin readme. You could shorten that up to do some arbitrary calculation on a field.

**Comment by jstirnaman** [comment]
@sanderson Just updated.

File: content/influxdb/cloud-dedicated/write-data/best-practices/optimize-writes.md [hunk 9]
```diff
@@ -100,13 +111,34 @@ For specific instructions, see the
...
 ```
+
 {{% /code-callout %}}
+{{% /code-placeholders %}}
+{{% /influxdb/custom-timestamps %}}
+
+Replace the following:
+
+- {{% code-placeholder-key %}}`DATABASE_NAME`{{% /code-placeholder-key %}}: the name of the [database](/influxdb/cloud-dedicated/admin/databases/) to write data to
+- **`token`**: a [token](/influxdb/cloud-dedicated/admin/tokens/) with _write_ access to the specified database.
```
Comment (sanderson) [inline content/influxdb/cloud-dedicated/write-data/best-practices/optimize-writes.md:139]:
```suggestion
- {{% code-placeholder-key %}}`DATABASE_TOKEN`{{% /code-placeholder-key %}}: a [token](/influxdb/cloud-dedicated/admin/tokens/) with _write_ access to the specified database.
```

File: content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md [hunk 10]
```diff
@@ -104,11 +114,34 @@ For specific instructions, see the
...
+
 {{% /code-callout %}}
+{{% /code-placeholders %}}
+{{% /influxdb/custom-timestamps %}}
+
+Replace the following:
+
+- {{% code-placeholder-key %}}`ORG_NAME`{{% /code-placeholder-key %}}: the name of your [organization](/influxdb/cloud-serverless/admin/organizations/)
+- {{% code-placeholder-key %}}`BUCKET_NAME`{{% /code-placeholder-key %}}: the name of the [bucket](/influxdb/cloud-serverless/admin/buckets/) to write data to
+- **`token`**: a [token](/influxdb/cloud-serverless/admin/tokens/) with _write_ access to the specified bucket.
```
Comment (sanderson) [inline content/influxdb/cloud-serverless/write-data/best-practices/optimize-writes.md:142]:
```suggestion
- {{% code-placeholder-key %}}`DATABASE_TOKEN`{{% /code-placeholder-key %}}: a [token](/influxdb/cloud-serverless/admin/tokens/) with _write_ access to the specified bucket.
```

File: content/influxdb/clustered/write-data/best-practices/optimize-writes.md [hunk 11]
```diff
@@ -112,11 +124,21 @@ mem,host=host2 used_percent=27.18294630 1641034800" | gzip > system.gzip \
...
 ```
+
 {{% /code-callout %}}
+{{% /code-placeholders %}}
 {{% /influxdb/custom-timestamps %}}
+
+Replace the following:
+
+- {{% code-placeholder-key %}}`DATABASE_NAME`{{% /code-placeholder-key %}}: the name of the [database](/influxdb/clustered/admin/databases/) to write data to
+- **`token`**: a [token](/influxdb/clustered/admin/tokens/) with _write_ access to the specified database.
```
Comment (sanderson) [inline content/influxdb/clustered/write-data/best-practices/optimize-writes.md:139]:
```suggestion
- {{% code-placeholder-key %}}`DATABASE_TOKEN`{{% /code-placeholder-key %}}: a [token](/influxdb/clustered/admin/tokens/) with _write_ access to the specified database.
```

**Review by sanderson — APPROVED** [review]
Looks great! Just some missed code placeholder keys that I added suggestions for. Thanks for this!

**Pull request merged**